### ISSUE TITLE
4 Byte BGP AS Numbers

### ISSUE THREAD

**Issue by neil2048** [vmware/powernsx#528]
Hi

Im automating an NSX-V install and the customer want to use 4 byte AS numbering, this was supported from NSX 6.3, how every when i code in PowerNSX the local or remote AS numbers, using the dot notation 65000.1824, this always apears as 65001 in NSX.  

is there a specific way to code this for PowerNSX?

Thanks

**Comment by neil2048** [comment]
Hi

Im using Several, cut some code out for clarity.. I have tried 650001824 as well as 65000.1824 and it round with the dot notation, and errors without  with :-
"Cannot validate argument on parameter 'RemoteAS'. The 664801824 argument is greater than the maximum allowed range of 65535. Supply an argument that is less than or equal to 65535 and then try the command again."

Get-Nsxlogicalrouter $($row[0]) | Get-NsxLogicalRouterRouting | Set-NsxLogicalRouterRouting -EnableBgp -ProtocolAddress $($row[18]) -ForwardingAddress $($row[8]) -LocalAS ($row[6])

Get-NsxLogicalRouter $($row[0]) | Get-NsxLogicalRouterRouting | New-NsxLogicalRouterBgpNeighbour -IpAddress $($row2[1]) -RemoteAS $($row2[2])

Get-NSXEdge $($row[0]) | Get-NsxEdgeRouting | Set-NsxEdgeRouting -EnableBgp -LocalAS $($row[6])
Get-NSXEdge $($row[0]) | Get-NsxEdgeRouting | New-NsxEdgeBgpNeighbour -RemoteAS $($row2[2])

**Comment by dcoghlan** [comment]
Hi @neil2048 

Thanks for logging the issue.

The problem is that the current validation for BGP AS numbers in PowerNSX is only checking for a value from 0-65535

With the introduction of the 4-Byte ASN, there are multiple options that NSX-v will now accept.

We will take a look at the code and see how we can implement better BGP AS validation across the board so that 4-Byte AS numbers are accepted.

However, to workaround the issue, you will likely need to comment out the validation code in the cmdlets you need to use for the time being.

If your feeling real adventurous, you could trying using `invoke-nsxrestmethod` and constructing your own XML.

Regards
Dale

**Comment by alagoutte** [comment]
Oh...

it will be more complicated... because API add a new "field" for 4 Bytes ASN : localASNumber (and RemoteASNumber for BGP Neighbours)
Need to add support of this new field...

**Comment by neil2048** [comment]
Hi

yes, i have a work around for the ESG's at least where i can call

$edge = Get-NSXEdge $($row[0])
$edge.features.routing.bgp.bgpNeighbours.bgpNeighbour[$peerid].remoteASNumber = $($row2[2]) 
$peerid++
$edge | Set-NSXEdge  -confirm:$false 

inside a foreach loop incrementing $peerid each iteration, this appears to work, however im still struggling to apply the same config to the LogicalRouters, Im just double checking the code on the LogicalRouter side. then Ill advise shortly if this workaround also works.

Thanks

**Comment by neil2048** [comment]
Yes the same code does work for the Get/Set-NSXLogicalRouter cmdlets also using $edge.features.routing.bgp.localASNumber =  $($row[6]) to set the local AS number to a 4 byte one.

**Comment by alagoutte** [comment]
Yes, it is the same "API" for NsxLogicialRouter and NSXEdge...

**Comment by Kyleset** [comment]
This looks like it has fallen by the wayside. I found a workaround for this. The cmdlet takes Int32, but will also take a string. Obviously the 4byte ASN wont fit in a Int32, but if you convert your Int64 decimal ASN to a string that is asdot notation, then the cmdlet will accept it. I created a function that will change it over. I used https://networklessons.com/bgp/bgp-4-byte-number to create a function. However, it sure would be nice to have it accept Int64.
